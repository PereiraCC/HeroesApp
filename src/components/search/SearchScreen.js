import { useNavigate, useLocation } from 'react-router-dom';
import queryString from 'query-string';

import { useForm } from '../../hook/useForm';
import { HeroCard } from '../hero/HeroCard';

import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);

    const [ { searchText }, handleInputChange ] = useForm( {
        searchText : q
    });

    const heroesFilted = getHeroesByName(q);

    const handleSearch = (e) => {
        e.preventDefault();
        navigate(`?q=${searchText}`)
    }

    return (
        <>
            <h1>Searches</h1>
            <hr />

            <div className='row'>

                <div className='col-5'>
                    <h4>Search</h4>
                    <hr />

                    <form onSubmit={ handleSearch }>
                        <input 
                            type="text"
                            placeholder='Search a hero'
                            className='form-control'
                            name='searchText'
                            autoComplete='off'    
                            onChange={ handleInputChange } 
                            value={ searchText }
                        />

                        <button 
                            className='btn btn-outline-primary mt-1 btn-block'
                            type='submit'
                        >
                            Search...
                        </button>
                    </form>
                </div>

                <div className='col-7'>
                    <h4> Results </h4>
                    <hr />

                    {
                        heroesFilted.map(hero => (
                            <HeroCard 
                                key={ hero.id }
                                { ...hero }
                            /> 
                        ))
                    }
                </div>

            </div>
        </>
    )
}
