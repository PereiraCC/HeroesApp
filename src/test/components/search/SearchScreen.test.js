import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}));

describe('<SearchScreen /> tests', () => {
   
    test('should show correctly with default values', () => {
        
        const wrapper = mount( 
            <MemoryRouter initialEntries={ ['/search'] }>
                <SearchScreen /> 
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-info').text().trim() ).toBe( 'Search a Hero' );

    });

    test('should show to Batman and the input with queryString value', () => {
       
        const wrapper = mount( 
            <MemoryRouter initialEntries={ ['/search?q=batman'] }>
                <SearchScreen /> 
            </MemoryRouter>
        );

        expect( wrapper.find('input').prop('value') ).toBe( 'batman' );
        expect( wrapper ).toMatchSnapshot();

    });

    test('should show error if hero not found', () => {
       
        const wrapper = mount( 
            <MemoryRouter initialEntries={ ['/search?q=batman123'] }>
                <SearchScreen /> 
            </MemoryRouter>
        );

        expect( wrapper.find('.alert-danger').text().trim() ).toBe( 'No results: batman123' );
        
    });

    test('should call navigate to new screen', () => {
       
        const wrapper = mount( 
            <MemoryRouter initialEntries={ ['/search'] }>
                <SearchScreen /> 
            </MemoryRouter>
        );

        wrapper.find( 'input' ).simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        });

        wrapper.find( 'form' ).prop('onSubmit')({
            preventDefault: () => {}
        });

        expect( mockNavigate ).toHaveBeenCalled();
        expect( mockNavigate ).toHaveBeenCalledWith('?q=batman');

    });
    
    
    
});
