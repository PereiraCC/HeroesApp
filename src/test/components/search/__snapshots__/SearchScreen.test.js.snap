// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchScreen /> tests should show correctly with default values 1`] = `
Array [
  <h1>
    Searches
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-1 btn-block"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a Hero
      </div>
    </div>
  </div>,
]
`;

exports[`<SearchScreen /> tests should show to Batman and the input with queryString value 1`] = `
Array [
  <h1>
    Searches
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-1 btn-block"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                   
                  Batman
                   
                </h5>
                <h5
                  className="card-text"
                >
                   
                  Bruce Wayne
                   
                </h5>
                <p
                  className="card-text"
                >
                  <small
                    className="tex-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  More...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
